safehttp/plugins/hsts/hsts.go

```diff
@@ -0,0 +1,82 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package hsts
+
+import (
+	"strconv"
+	"strings"
+
+	"github.com/google/go-safeweb/safehttp"
+)
+
+// Plugin implements automatic HSTS functionality.
+type Plugin struct {
+	// The time, in seconds, that the browser should remember
+	// that a site is only to be accessed using HTTPS.
+	MaxAge uint64
+
+	// This field controls the includeSubDomains directive.
+	// When DisableIncludeSubDomains is false, all subdomains
+	// of the domain where this service is hosted will also be added
+	// to the browsers HSTS list.
+	DisableIncludeSubDomains bool
+
+	// This field controls the preload directive.
+	// This should only be enabled if this site should be
+	// added to the browser HSTS preload list, which is supported
+	// by all major browsers. See https://hstspreload.org/ for
+	// more info.
+	Preload bool
+
+	// If this server is behind a proxy that terminates HTTPS
+	// traffic then this should be enabled. If this is enabled
+	// then the plugin will always send the Strict-Transport-Security
+	// header and will not redirect HTTP traffic to HTTPS traffic.
+	BehindProxy bool
+}
+
+// NewPlugin creates a new HSTS plugin with safe defaults.
+func NewPlugin() Plugin {
+	return Plugin{MaxAge: 63072000} // two years in seconds
+}
+
+// Before should be executed before the request is sent to the handler.
+// The function redirects HTTP requests to HTTPS. When HTTPS traffic
+// is received the Strict-Transport-Security header is applied to the
+// response.
+func (p *Plugin) Before(w safehttp.ResponseWriter, r *safehttp.IncomingRequest) safehttp.Result {
+	if !p.BehindProxy && r.TLS == nil {
+		r.URL.Scheme = "https"
```

Comment:
I'd not do this. Imagine a scenario:

You've installed two plugins, `foo.Plugin` and `hsts.Plugin`. `foo` runs before `hsts`.

- `foo.Before` runs, then `hsts.Before`
- `hsts.After` runs (since `it has run `Before`)
- `foo.After` also runs, because it has already run it's `Before`
- `foo.After` reads the `IncomingRequest.URL`, because it logs all URLs visited when non-200 status was produced as a response

Reply:
Ah, I thought that if a plugin interrupts during the `Before` phase then the `After` phase is not run.